// @stryker-mutator/typescript-checker/dist/src/fs/script-file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScriptFile = void 0;
const tslib_1 = require("tslib");
const typescript_1 = (0, tslib_1.__importDefault)(require("typescript"));
class ScriptFile {
    constructor(content, fileName, modifiedTime = new Date()) {
        this.content = content;
        this.fileName = fileName;
        this.modifiedTime = modifiedTime;
        this.originalContent = content;
    }
    write(content) {
        this.content = content;
        this.touch();
    }
    mutate(mutant) {
        this.guardMutationIsWatched();
        const start = this.getOffset(mutant.location.start);
        const end = this.getOffset(mutant.location.end);
        this.content = `${this.originalContent.substr(0, start)}${mutant.replacement}${this.originalContent.substr(end)}`;
        this.touch();
    }
    getOffset(pos) {
        if (!this.sourceFile) {
            this.sourceFile = typescript_1.default.createSourceFile(this.fileName, this.content, typescript_1.default.ScriptTarget.Latest, false, undefined);
        }
        return this.sourceFile.getPositionOfLineAndCharacter(pos.line, pos.column);
    }
    resetMutant() {
        this.guardMutationIsWatched();
        this.content = this.originalContent;
        this.touch();
    }
    guardMutationIsWatched() {
        if (!this.watcher) {
            throw new Error(`Tried to check file "${this.fileName}" (which is part of your typescript project), but no watcher is registered for it. Changes would go unnoticed. This probably means that you need to expand the files that are included in your project.`);
        }
    }
    touch() {
        var _a;
        this.modifiedTime = new Date();
        (_a = this.watcher) === null || _a === void 0 ? void 0 : _a.call(this, this.fileName, typescript_1.default.FileWatcherEventKind.Changed);
    }
}
exports.ScriptFile = ScriptFile;
